src/_guides/language/effective-dart/design.md

```diff
@@ -1005,79 +1091,260 @@ var names = people.map((Person person) {
 });
 {% endprettify %}
 
+In rare cases, the surrounding context is not precise enough to provide a type
+for one or more of the function's parameters. In those cases, you may need to
+annotate.
+
 
-### AVOID annotating with `dynamic` when not required.
+### PREFER annotating with `dynamic` even when not required.
 
-{% comment %}
-update-for-dart-2.0
-https://github.com/dart-lang/site-www/pull/480#discussion_r157355593
-{% endcomment %}
+When you omit a type annotation, *most* of the time, Dart infers a correct
+static type for you. That word "most" is the alarming part. Sometimes, inference
+fails and silently uses `dynamic` instead. If you didn't realize that inference
+failed, you now have code that *looks* statically safe but actually has no
+type checking whatsoever.
 
-In most places in Dart, a type annotation can be omitted, in which case the type
-will automatically be `dynamic`. Thus, omitting the type annotation entirely is
-semantically equivalent but more terse.
+Sometimes, `dynamic` is what you want. Dart does support a dynamic type for good
+reasons. But a casual reader of your code who sees an annotation is missing has
+no way of knowing if you intended it to be `dynamic`, expected inference to fill
+in some other type, or you simply forgot to write the annotation.
+
+Using `dynamic` explicitly when that is the type you want makes that intention
+clear.
 
 {:.good-style}
-<?code-excerpt "misc/lib/effective_dart/design_good.dart (avoid-dynamic)"?>
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (prefer-dynamic)"?>
 {% prettify dart %}
-lookUpOrDefault(String name, Map map, defaultValue) {
+T lookUp<T>(String name, Map<String, T> map, T defaultValue) {
```

Comment:
The emphasis in this code excerpt are the type annotations so _maybe_ the body can be elided. If we keep the body, how about making it more null-operator smart (in both the good and the bad examples)?

```dart
  T lookUp<T>(String name, Map<String, T> map, T defaultValue) =>
      map[name] ?? defaultValue;
```

Reply:
👍 Done.